### PR TITLE
Ball & Chain + Fissure Spawner + Fix #242

### LINKED ISSUE #242: Incorrect subtype for water flow (right) and water flow (up)
They're mixed up atm, up should be subtype 1 and right should be subtype 2

### PR DESCRIPTION
**Pull request by Rchardon**

### CODE REVIEW

File: BasementRenovator.py [hunk 1]
```diff
@@ -1273,6 +1273,10 @@ def getEntityInfo(self, t, variant, subtype):
                 en = entityXML.find(
                     f"entity[@ID='{t}'][@Subtype='{subtype}'][@Variant='{variant}']"
                 )
+                if t in (893, 3001):
```
Comment (Zamiell) [inline BasementRenovator.py:1276]: no magic numbers! pls fix

File: BasementRenovator.py [hunk 2]
```diff
@@ -1387,14 +1391,16 @@ def updateTooltip(self):
             tooltipStr += "\nType is outside the valid range of 0 - 999! This will not load properly in-game!"
         if e.Variant >= 4096:
             tooltipStr += "\nVariant is outside the valid range of 0 - 4095!"
-        if e.Subtype >= 255:
+        if e.Type not in (893, 3001) and e.Subtype >= 255:
```
Comment (Zamiell) [inline BasementRenovator.py:1394]: should be a previously defined array, don't repeat yourself

File: BasementRenovator.py [hunk 3]
```diff
@@ -1387,14 +1391,16 @@ def updateTooltip(self):
...
             tooltipStr += "\nSubtype is outside the valid range of 0 - 255!"
         if e.invalid:
             tooltipStr += "\nMissing entities2.xml entry! Trying to spawn this WILL CRASH THE GAME!!"
         if not e.known:
             tooltipStr += (
                 "\nMissing BR entry! Trying to spawn this entity might CRASH THE GAME!!"
             )
+        if e.Type in (893, 3001):
```
Comment (Zamiell) [inline BasementRenovator.py:1402]: same here

File: resources/EntitiesRepentance.xml [hunk 4]
```diff
@@ -392,6 +392,7 @@
 	<entity BaseHP="15" Champion="1" Group="Hosts" ID="27" Image="resources/Entities/27.3.0 - Hard Host.png" Kind="Enemies" Name="Hard Host" Variant="3" Subtype="0" />
 	<entity BaseHP="20" Champion="1" Group="Hosts" ID="859" Image="resources/Entities/859.0.0 - Floating Host.png" Kind="Enemies" Name="Floating Host" Variant="0" Subtype="0" />
 	<entity Group="Hosts" Kind="Enemies" Image="resources/Entities/300.0.0 - MushroomMan.png" BaseHP="23" Champion="1" ID="300" Name="Mushroom" Variant="0" Subtype="0" />
+	<entity BaseHP="100" Group="Hosts" ID="3001" Image="resources/Entities/3001.0.0 - Fissure spawner.png" Kind="Enemies" Name="Fissure spawner" Variant="0" Subtype="0" IsGrid="1"/>
```
Comment (Zamiell) [inline resources/EntitiesRepentance.xml:395]: shouldn't it be "Fissure Spawner" ? i think that would be more consistent

File: BasementRenovator.py [hunk 5]
```diff
@@ -1920,6 +1939,112 @@ def hideWeightPopup(self):
...
+        if self.entity.Type == 893:
+            # Rotation
+            Rotation = QWidgetAction(menu)
+            r = QComboBox()
+            r.addItem("Clockwise")
+            r.addItem("Counter clockwise")
+            r.setCurrentIndex(self.property1)
+
+            Rotation.setDefaultWidget(r)
+            r.currentIndexChanged.connect(self.changeProperty1)
```
Comment (budjmt) [inline BasementRenovator.py:2005]:
```suggestion
            r.setCurrentIndex(self.properties[0])

            Rotation.setDefaultWidget(r)
            r.currentIndexChanged.connect(lambda x => self.changeProperty(0, x))
```

File: BasementRenovator.py [hunk 6]
```diff
@@ -1920,6 +1939,112 @@ def hideWeightPopup(self):
...
+            d.setPrefix("Angle - ")
+        d.setValue(self.property3)
+
+        Distance.setDefaultWidget(d)
+        d.valueChanged.connect(self.changeProperty3)
+        menu.addAction(Distance)
+
+        # End it
+        menu.exec(self.list.mapToGlobal(pos))
+
```
Comment (budjmt) [inline BasementRenovator.py:2046]: regarding the issue with the context menu reappearing everywhere, why not make this whole class extend QMenu and exec it from the mouse handler? that might make it work

**Review by budjmt — CHANGES_REQUESTED** [review]
Nice attention to Qt stuff! I'll think about how this system can be made more generic from an xml perspective

File: BasementRenovator.py [hunk 7]
```diff
@@ -1874,8 +1885,15 @@ def remove(self):
...
-        QGraphicsItem.mouseReleaseEvent(self, event)
+        e = self.entity
+        if (
+            event.button() == Qt.MiddleButton
+            and e.Type in Entity.ENTITIES_WITH_PROPERTIES
+        ):
+            EntityMenu(e)
+        else:
+            self.hideWeightPopup()
+            QGraphicsItem.mouseReleaseEvent(self, event)
```
Comment (budjmt) [inline BasementRenovator.py:1896]: I think we want to process this code regardless, not sure if that causes issues
```suggestion
        self.hideWeightPopup()
        QGraphicsItem.mouseReleaseEvent(self, event)
```

**Review by budjmt — CHANGES_REQUESTED** [review]

**Comment by Zamiell** [comment]
gamon is this ready for PR now?

**Comment by Rchardon** [comment]
yes you can go

**Comment by Zamiell** [comment]
lets wait for @budjmt to review it

**Review by budjmt — APPROVED** [review]

**Pull request merged**